pkg/controller/database/database_controller_test.go

```diff
@@ -0,0 +1,199 @@
+func TestDbControllerTest(t *testing.T) {
+	// create envtest instance, point to binaries
+	// and load CRDs
+	env := &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join("..", "..", "..", "config", "crds", "v1"),
+		},
+		BinaryAssetsDirectory: filepath.Join(os.TempDir(), "kubebuilder", "bin"),
+	}
+
+	// start the kube api server and etcd
+	cfg, err := env.Start()
+	if err != nil {
+		t.Fatalf("Failed to start environment: %v", err)
+	}
+
+	if cfg == nil {
+		t.Fatal("Config was nil")
+	}
+
+	mgr, err := manager.New(cfg, manager.Options{})
+	if err != nil {
+		t.Fatalf("Failed to create manager: %v", err)
+	}
+
+	// add types to Scheme
+	err = apis.AddToScheme(mgr.GetScheme())
+	if err != nil {
+		t.Fatalf("Failed to setup Scheme: %v", err)
+	}
+
+	// add the Controller to the Manager
+	err = Add(mgr, "", "")
+	if err != nil {
+		t.Fatalf("Failed to add controller to manager: %v", err)
+	}
+
+	// start the controller manager in a goroutine so it doesn't
+	// block envtest cleanup
+	go func() {
+		err = mgr.Start(signals.SetupSignalHandler())
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	client := mgr.GetClient()
+
+	tests := []struct {
+		name           string
+		namespace      corev1.Namespace
+		expectedLabels map[string]string
+		customLabels   map[string]string
+	}{
+		{
+			name: "creates_statefulset_with_default_labels",
+			expectedLabels: map[string]string{
+				"control-plane": "schemahero",
+				"database":      "test",
+			},
+		},
+		{
+			name: "creates_statefulset_with_default_and_custom_labels",
+			customLabels: map[string]string{
+				"custom": "label",
+			},
+			expectedLabels: map[string]string{
+				"control-plane": "schemahero",
+				"database":      "test",
+				"custom":        "label",
+			},
+		},
+	}
+
+	// wrapping the subtests in a t.Run so we wait till they're all
+	// done before bringing down envtest
+	t.Run("group", func(t *testing.T) {
+		for _, test := range tests {
+			test := test
+
+			t.Run(test.name, func(t *testing.T) {
+				t.Parallel()
+
+				// create a new Namespace for each run to
+				// create Objects in to avoid collisions
+				ns, err := CreateNamespace(client)
+				assert.NoError(t, err)
+
+				db := databasesv1alpha4.Database{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test",
+						Namespace: ns.Name,
+					},
+					Spec: databasesv1alpha4.DatabaseSpec{
+						Template: &databasesv1alpha4.DatabaseTemplate{
+							ObjectMeta: metav1.ObjectMeta{
+								Labels: test.customLabels,
+							},
+						},
+					},
+				}
+
+				err = client.Create(context.Background(), &db)
+				assert.NoError(t, err)
+
+				ss, err := GetStatefulSet(client, ns.Name, fmt.Sprintf("%s-controller", db.Name), 5, 1)
+				assert.NoError(t, err)
+
+				actualLabels := ss.Spec.Template.ObjectMeta.Labels
+				assert.True(t, HasLabels(test.expectedLabels, actualLabels), "Wanted: %v\ngot: %v", test.expectedLabels, actualLabels)
```

Comment:
Can we just use `require.Equal` or `assert.Equal` instead of the custom HasLabels function? Unless we're intentionally allowing the addition of further labels

Reply:
Yeah wanted to future proof against any other labels that might get added by default? The logic is supposed to be that "my custom labels get added to the mix" and not "these are the only labels that should exist"